composition-js/src/merging/merge.ts

```diff
@@ -803,19 +821,218 @@ class Merger {
   }
 
   private isShareable(sourceIdx: number, field: FieldDefinition<any>): boolean {
     return this.metadata(sourceIdx).isFieldShareable(field);
   }
 
-  private mergeField(sources: (FieldDefinition<any> | undefined)[], dest: FieldDefinition<any>) {
+  private isFieldKey(sourceIdx: number, field: FieldDefinition<any> | undefined): boolean {
+    if (!field) {
+      return false;
+    }
+    return this.metadata(sourceIdx).isFieldKey(field);
+  }
+
+  private getOverrideDirective(sourceIdx: number, field: FieldDefinition<any>): Directive<any> | undefined {
+    // Check the directive on the field, then on the enclosing type.
+    const metadata = this.metadata(sourceIdx);
+    const overrideDirective = metadata.isFed2Schema() ? metadata.overrideDirective() : undefined;
+    const allFieldOverrides = overrideDirective ? field.appliedDirectivesOf(overrideDirective) : [];
+    return allFieldOverrides.find(d => d.ofExtension() === field.ofExtension());
+
+    // TODO: commenting out since we don't support override on type just yet. Will be added back shortly
+    /*
+    const type = field.parent;
+    assert(type, 'Field must have a parent');
+    const allTypeOverrides: Directive<any>[] = overrideDirective ? type.appliedDirectivesOf(overrideDirective) : [];
+    const overrideOnType = allTypeOverrides.find((d) => d.ofExtension() === type.ofExtension);
+
+    // if both directives are present, raise an error
+    if (overrideOnField && overrideOnType) {
+      this.errors.push(ERRORS.OVERRIDE_ON_BOTH_FIELD_AND_TYPE.err({
+        message: `Field "${field.coordinate}" on subgraph "${this.names[sourceIdx]}" is marked with @override directive on both the field and the type`,
+      }));
+    } else if (overrideOnField || overrideOnType) {
+      return overrideOnField ?? overrideOnType; // only one of them is not undefined, but this will return the one we want
+    }
+    return undefined;
+    */
+  }
+
+  private overrideHasIncompatibleFields({
+    idx,
+    field,
+    subgraphName,
+    fromIdx,
+    fromField,
+    fromSubgraphName
+  }: {
+    idx: number;
+    field: FieldDefinition<any> | undefined;
+    subgraphName: string;
+    fromIdx: number;
+    fromField: FieldDefinition<any> | undefined;
+    fromSubgraphName: string;
+  }): { result: boolean, conflictingDirective?: string, subgraph?: string } {
+    // helper function to let us know if a given directive exists on a field
+    const directiveOnField = (field: FieldDefinition<any> | undefined, def: DirectiveDefinition<any> | undefined): boolean => {
+      if (!def || !field) {
+        return false;
+      }
+      const directives = field.appliedDirectivesOf(def);
+      const directiveOnField = directives.find(d => d.ofExtension() === field.ofExtension());
+      const directivesOnType: Directive<any>[] = field.parent.appliedDirectivesOf(def);
+      const directiveOnType = directivesOnType.find(d => d.ofExtension() === field.parent.ofExtension);
```

Comment:
Same remark that in `getOverrideDirective` regarding the extension check for field directives. And more generally, this helper is only called for provides and requires and neither of them can be applied on types, so I think that helper can be removed and `directiveOnField(fromField, fromMetadata.providesDirective())` could be replaced by `fromField.hasAppliedDirective(fromMetadata.providesDirectives())`.

As an aside, I agree with not allowing overriding a field with a provides/requires (at least for now), but it would be awesome to add quick tests for those.